cmd/test-server/main.go

```diff
@@ -215,3 +276,91 @@ func (p *TestServer) runTestServer(ctx context.Context, o *TestServerRunOptions)
 	}
 	return stop, nil
 }
+
+func (p *TestServer) runFlakyTestServer(ctx context.Context, o *TestServerRunOptions) (StopFunc, error) {
+	muxHandler := http.NewServeMux()
+	muxHandler.HandleFunc("/success", returnSuccess)
+	muxHandler.HandleFunc("/sleep", sleepReturnSuccess)
+	muxHandler.HandleFunc("/error", returnError)
+	muxHandler.HandleFunc("/close", closeNoResponse)
+
+	address := fmt.Sprintf("127.0.0.1:%d", o.flakyPort)
+	ln, err := newFlakyListener("tcp", address, convertFlakeMode(o.flakeType))
+	if err != nil {
+		klog.Warningf("failed to create new flaky listener: %v\n", err)
+	}
+	go func() {
+		err = http.Serve(ln, muxHandler)
+		if err != nil {
+			klog.Warningf("HTTP test server received %v.\n", err)
+		}
+	}()
+
+	stop := func() {
+		err := ln.Close()
+		if err != nil {
+			klog.ErrorS(err, "failed to shutdown flaky server")
+		}
+	}
+	return stop, nil
+}
+
+// FlakyListener currently ignores 1/2 incoming requests.
+// should extend to handle other types of connection errors.
+type FlakyListener struct {
+	inner net.Listener
+	flipper bool
+	mode flakeMode
+}
+
+func newFlakyListener(network, address string, mode flakeMode) (net.Listener, error) {
+	ln, err := net.Listen("tcp", address)
+	if err != nil {
+		return nil, err
+	}
+	return &FlakyListener{
+		inner: ln,
+		flipper: false,
+		mode: mode,
+	}, nil
+}
+
+// Accept waits for and returns the next connection to the listener.
+func (f* FlakyListener) Accept() (net.Conn, error) {
+	c, err := f.inner.Accept()
```

Comment:
I'm not familiar with how Accept() works. Maybe when inner.Accept() is called, from the client's perspective, the dial has succeeded, so we never trigger the dial timeout in the proxy agent.

The fact that flakyListener.Accept returns a net.Conn half of the time perhaps only affects how the following HTTP requests are handled but doesn't affect the dial.